src/modelframe.jl

```diff
@@ -1,261 +1,178 @@
+    model_cols(f.lhs, columntable(data))
 end
 
 
+StatsBase.modelmatrix(mf::ModelFrame; data=mf.data) = model_cols(mf.f.rhs, data)
+StatsBase.response(mf::ModelFrame; data=mf.data) = model_cols(mf.f.lhs, data)
+
+StatsBase.coefnames(mf::ModelFrame) = vectorize(coefnames(mf.f.rhs))
+
 """
-    termnames(term::Symbol, col)
-Returns a vector of strings with the names of the coefficients
-associated with a term.  If the column corresponding to the term
-is not categorical, a one-element vector is returned.
-"""
-termnames(term::Symbol, col) = [string(term)]
-function termnames(term::Symbol, mf::ModelFrame; non_redundant::Bool = false)
-    if haskey(mf.contrasts, term)
-        termnames(term, mf.df[term],
-                  non_redundant ?
-                  convert(ContrastsMatrix{FullDummyCoding}, mf.contrasts[term]) :
-                  mf.contrasts[term])
-    else
-        termnames(term, mf.df[term])
-    end
-end
+    setcontrasts!(mf::ModelFrame; kwargs...)
+    setcontrasts!(mf::ModelFrame, contrasts::Dict{Symbol})
 
-termnames(term::Symbol, col::Any, contrast::ContrastsMatrix) =
-    ["$term: $name" for name in contrast.termnames]
+Update the contrasts used for coding categorical variables in
+[`ModelFrame`](@ref) in place.  This is accomplished by computing a new schema
+based on the provided contrasts and the `ModelFrame`'s data, and applying it to
+the `ModelFrame`'s `FormulaTerm`.
 
+Note that only the `ModelFrame` itself is mutated: because `AbstractTerm`s are
+immutable, any changes will produce a copy.
 
-function expandtermnames(term::Vector)
-    if length(term) == 1
-        return term[1]
-    else
-        return foldr((a,b) -> vec([string(lev1, " & ", lev2) for
-                                   lev1 in a,
-                                   lev2 in b]),
-                     term)
+"""
+setcontrasts!(mf::ModelFrame; kwargs...) = setcontrasts!(mf, Dict(kwargs))
+function setcontrasts!(mf::ModelFrame, contrasts::Dict{Symbol})
+    # TODO: don't consume the whole table again if it's not needed
+    new_schema = schema([term(k) for k in keys(contrasts)], mf.data, contrasts)
+
+    # warn of no-op for keys that dont correspond to known terms from old schema
+    unknown_keys = [k for k in keys(new_schema) if !haskey(mf.schema, k)]
+    if !isempty(unknown_keys)
+        unknown_keys_str = join(unknown_keys, ", ", " and ")
+        @warn "setcontrasts! for terms " * unknown_keys_str *
+            " has no effect since they are not found in original schema"
     end
+
+    # apply only the re-mapped terms
+    mf.f = apply_schema(mf.f, new_schema, mf.model)
+    merge!(mf.schema, new_schema)
+    mf
 end
 
 
 """
-    coefnames(mf::ModelFrame)
-Returns a vector of coefficient names constructed from the Terms
-member and the types of the evaluation columns.
-"""
-function StatsBase.coefnames(mf::ModelFrame)
-    terms = droprandomeffects(dropresponse!(mf.terms))
+Convert a `ModelFrame` into a numeric matrix suitable for modeling
```

Comment:
I was avoiding this because it looks strange in the generated docs (since documenter prints a header with the name of the type anyway).  But I suppose that for docstrings in the REPL it would still be useful to have?